### PR TITLE
modules: update Azure deps to latest

### PR DESCRIPTION
**Pull request by vancluever**
The dependencies here are dated and are causing conflicts with the
ACME provider, namely the version of the top-level autorest package.

This explicitly updates the Azure SDK and autorest packages, with the
separately versioned sub-packages being added automatically.

### CODE REVIEW

**Comment by vancluever** [comment]
Apologies - I had meant to push this to my own fork but it ended up here. 😕 

Leaving it open so we don't have to re-open the PR.

File: go.mod [hunk 1]
```diff
@@ -2,8 +2,13 @@ module github.com/hashicorp/terraform
 
 require (
 	cloud.google.com/go v0.36.0
-	github.com/Azure/azure-sdk-for-go v21.3.0+incompatible
-	github.com/Azure/go-autorest v10.15.4+incompatible
+	github.com/Azure/azure-sdk-for-go v31.2.0+incompatible
+	github.com/Azure/go-autorest v12.3.0+incompatible // indirect
+	github.com/Azure/go-autorest/autorest v0.5.0
```
Comment (tamalsaha) [inline go.mod]: You probably don't need dependency on both 

```
	github.com/Azure/go-autorest v12.3.0+incompatible // indirect
	github.com/Azure/go-autorest/autorest v0.5.0
```

Please see here: https://github.com/Azure/go-autorest#using-with-go-modules

Comment (vancluever) [reply go.mod]: @tamalsasha the project does not build without this. Without the root-level dependency the build process will attempt to fall back to an older version that causes conflicts.

You can try it yourself with a clean repo and then just attempting to update `github.com/Azure/azure-sdk-for-go` and then running a `go mod tidy`.

Comment (vancluever) [reply go.mod]: Even doing `go get github.com/Azure/azure-sdk-for-go@v31.2.0 github.com/Azure/go-autorest@v12.3.0` followed by `go mod tidy` will attempt to add a bad version back, so you need to re-add `v12.3.0` afterwards.

Comment (vancluever) [reply go.mod]: I didn't see the replace line though, so I'll give that a shot too!

Comment (vancluever) [reply go.mod]: @tamalsasha yeah thanks for this. The big thing was the `replace` line for sure which I will get added to this PR.

Thanks!

Comment (tamalsaha) [reply go.mod]: Glad it worked!

**Comment by pselle** [comment]
@vancluever Could you resolve conflicts on this and #22248 and then request review?

**Comment by vancluever** [comment]
@pselle done!

**Comment by tamalsaha** [comment]
FWIW, LGTM.

**Pull request merged**